### PR TITLE
DM-33817: Enable mypy checking

### PR DESCRIPTION
**Pull request by andy-slac**
Added mypy.ini and updated the code for missing type annotations. Added
github actions for mypy.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

**Review by timj — APPROVED** [review]
Looks okay to me.

File: python/lsst/ctrl/mpexec/singleQuantumExecutor.py [hunk 1]
```diff
@@ -136,22 +141,23 @@ def execute(self, taskDef, quantum, butler):
...
+    def _execute(self, taskDef: TaskDef, quantum: Quantum, butler: Butler) -> Quantum:
         """Internal implementation of execute()"""
         startTime = time.time()
 
         with self.captureLogging(taskDef, quantum, butler) as captureLog:
 
             # Save detailed resource usage before task start to metadata.
             quantumMetadata = _TASK_METADATA_TYPE()
-            logInfo(None, "prep", metadata=quantumMetadata)
+            logInfo(None, "prep", metadata=quantumMetadata)  # type: ignore
```
[the remark this answers is not in the document]
Comment (andy-slac) [reply python/lsst/ctrl/mpexec/singleQuantumExecutor.py:160]: `quantumMetadata` is a `TaskMetadata` instance and `logInfo` expects `MutableMapping` but apparently it also works with `TaskMetadata`. Should I change this code to make something else than `TaskMetadata`?

File: python/lsst/ctrl/mpexec/dataid_match.py [hunk 2]
```diff
@@ -22,11 +22,11 @@
 __all__ = ["DataIdMatch"]
 
 import operator
-from typing import Any, List, Optional, Tuple
+from typing import Any, Callable, List, Optional, Tuple
 
 import astropy.time
 from lsst.daf.butler import DataId
-from lsst.daf.butler.registry.queries.expressions import Node, ParserYacc, TreeVisitor
+from lsst.daf.butler.registry.queries.expressions import Node, ParserYacc, TreeVisitor  # type: ignore
```
[the remark this answers is not in the document]
Comment (andy-slac) [reply python/lsst/ctrl/mpexec/dataid_match.py:29]: For some reason mypy says that `ParserYacc` is not in that module. Import works fine and I tried to figure out why mypy is unhappy but could not.

File: python/lsst/ctrl/mpexec/singleQuantumExecutor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (timj) [reply python/lsst/ctrl/mpexec/singleQuantumExecutor.py:160]: Ah. `TaskMetadata` is the right thing and the intent was for it to be a mutable mapping. It implements the interface but I had forgotten that pydantic models don't look like mutable mapping even if they act like it.

File: setup.cfg [hunk 3]
```diff
@@ -39,6 +39,9 @@ tests_require =
 [options.packages.find]
 where=python
 
+[options.package_data]
+lsst.pipe.base = py.typed
```
Comment (timj) [inline setup.cfg:43]: Probably should be `lsst.ctrl.mpexec = py.typed`.... It can be fixed the next time someone works on the package.

Comment (andy-slac) [reply setup.cfg:43]: 🤦‍♂️ Sorry, pasting without thinking. I'm fixing it already on my next ticket.

**Pull request merged**